source/Octostache/Templates/Functions/TextManipulationFunction.cs

```diff
@@ -137,6 +138,96 @@ internal class TextManipulationFunction
             }
         }
 
+        public static string? Indent(string? argument, string[] options)
+        {
+            if (argument == null)
+            {
+                return null;
+            }
+
+            if (argument.Length == 0)
+            {
+                // No content, no indenting
+                return string.Empty;
+            }
+
+            var indentOptions = new IndentOptions(options);
+
+            if (!indentOptions.IsValid)
+            {
+                return null;
+            }
+
+            return indentOptions.InitialIndent + argument.Replace("\n", "\n" + indentOptions.SubsequentIndent);
+        }
+
+        private class IndentOptions
+        {
+            private static Regex dualSizeEx = new Regex(@"^((\d{1,3})?/)?(\d{1,3})$", RegexOptions.Compiled);
```

Comment:
Would it make sense to use a common approach for supplying the number of spaces to indent and the strings to indent? For example `#{Blah | Indent 4}` and `#{Blah | Indent "    "}` are the same, and `#{Blah | Indent 0 4}` and `#{Blah | "" "    "}` are the same.

Reply:
On second thought, given the ability to append to a string, can we get away with not having two parameters? It looks like the golang indent function will only indent after a line break: https://github.com/Masterminds/sprig/blob/3ac42c7bc5e4be6aa534e036fb19dde4a996da2e/strings.go#L109, and maybe we can do the same.

So if you need to change the indent before the first line, you would append the custom spacing: `#{MySuffix| Prepend "    " | Indent 8}"

Reply:
There is definitely some overlap between `Indent` and `Prepend`/`Append`, which is really obvious if you set the subsequent indent to nothing (effectively turning it into a prepend).

I think there is improved usability by having the logic wrapped up in a function to create a higher-level abstraction. I *could* do the following today to get a list:

```
#{if variable}- #{variable | Replace "
" "
- "}#{/if}
```

Or it could be:

```
#{variable | Indent "- "}
```

Which I think is more readable. Plus, because it is encapsulated you can chain the result:

```
#{variable | Indent "- " | MarkdownToHtml}
```

I settled on the "initial/subsequent" syntax because it let you express the basic pattern in a single option, which also made it wholly distinguishable from the two option variant which treats the values as the strings to indent by. This meant that there was an escape hatch if the user ever wanted to indent using a pattern that would be picked up as an indent size by specifying the value twice.

As for Go, it does add the padding to the beginning of the string:

```go
func indent(spaces int, v string) string {
	pad := strings.Repeat(" ", spaces)
	return pad + strings.Replace(v, "\n", "\n"+pad, -1)
}
```

It includes the padding for empty strings though. I referenced the Python implementation as it considers more complex indentation.

Reply:
👍 Sounds good. I think the docs will need to call out the edge cases around padding and empty strings (I had to look twice and still got mixed up with Helm/Go), but lets do it.

Reply:
Yes, I'll definitely be making sure things are clear in the documentation.